plugins/out_datadog/datadog.c

```diff
@@ -0,0 +1,279 @@
+    /* Format the time, use milliseconds precision not nanoseconds */
+    timestamp = tms->tm.tv_sec * 1000;
+    timestamp += tms->tm.tv_nsec/1000000;
+    /* round up if necessary */
+    if (tms->tm.tv_nsec % 1000000 >= 500000) {
+        ++timestamp;
+    }
+    return timestamp;
+}
+
+static void dd_msgpack_pack_key_value_str(msgpack_packer* mp_pck,
+                                          const char *key, size_t key_size,
+                                          const char *val, size_t val_size)
+{
+    msgpack_pack_str(mp_pck, key_size);
+    msgpack_pack_str_body(mp_pck, key, key_size);
+    msgpack_pack_str(mp_pck, val_size);
+    msgpack_pack_str_body(mp_pck,val, val_size);
+}
+
+static int datadog_format(const void *data, size_t bytes,
+                          const char *tag, int tag_len,
+                          char **out_data, size_t *out_size,
+                          struct flb_out_datadog *ctx)
+{
+    /* for msgpack global structs */
+    size_t off = 0;   
+    int array_size = 0;
+    msgpack_unpacked result;
+    msgpack_sbuffer mp_sbuf;
+    msgpack_packer mp_pck;
+    /* for sub msgpack objs */
+    int map_size;
+    struct flb_time tms;
+    int64_t timestamp;
+    msgpack_object *obj;
+    msgpack_object map;
+    msgpack_object root;
+    msgpack_object k;
+    msgpack_object v;
+    /* output buffer */
+    flb_sds_t out_buf;
+
+    /* Count number of records */
+    msgpack_unpacked_init(&result);
+    while (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        array_size++;
+    }
+    msgpack_unpacked_destroy(&result);
+    msgpack_unpacked_init(&result);
+
+    /* Create temporal msgpack buffer */
+    msgpack_sbuffer_init(&mp_sbuf);
+    msgpack_packer_init(&mp_pck, &mp_sbuf, msgpack_sbuffer_write);
+
+    /* Prepare array for all entries */
+    msgpack_pack_array(&mp_pck, array_size);
+
+    off = 0;
+    while (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        root = result.data;
+
+        /* Get timestamp and object */
+        flb_time_pop_from_msgpack(&tms, &result, &obj);
+        timestamp = timestamp_format(&tms);
+
+        map = root.via.array.ptr[1];
+        map_size = map.via.map.size;
+
+        /* build new object(map) with additional space for datadog entries */
+        msgpack_pack_map(&mp_pck, ctx->nb_additional_entries + map_size);
+
+        /* timestamp */
+        msgpack_pack_str(&mp_pck, flb_sds_len(ctx->json_date_key));
+        msgpack_pack_str_body(&mp_pck,
+                              ctx->json_date_key,
+                              flb_sds_len(ctx->json_date_key));
+        msgpack_pack_int64(&mp_pck, timestamp);
+
+        /* include_tag_key */
+        if (ctx->include_tag_key == FLB_TRUE) {
+            dd_msgpack_pack_key_value_str(&mp_pck,
+                                          ctx->tag_key, flb_sds_len(ctx->tag_key),
+                                          tag, sizeof(tag_len)-1);
```

Comment:
sizeof(tag_len) will become the data type size, tag_len should be fine

Reply:
ouch that was bad... (wrong copy/past)